cmd/datasetIngestor/main_test.go

```diff
@@ -70,232 +71,122 @@ func TestMainOutput(t *testing.T) {
-			passing = false
-		}
-		if addCaption != "" {
-			t.Logf("addCaption's default should be \"\" but it's \"%s\"", addCaption)
-			passing = false
-		}
-		if showVersion {
-			t.Logf("showVersion's default should be false but it's true")
-			passing = false
-		}
-		if !passing {
-			t.Fail()
-		}
+func TestMainFlags(t *testing.T) {
+	// test cases
+	tests := []struct {
+		name        string
+		boolFlags   map[string]bool
+		stringFlags map[string]string
+		intFlags    map[string]int
+		args        []string
+	}{
+		{
+			name: "Test without flags",
+			boolFlags: map[string]bool{
+				"ingest":              false,
+				"testenv":             false,
+				"devenv":              false,
+				"localenv":            false,
+				"tunnelenv":           false,
+				"noninteractive":      false,
+				"copy":                false,
+				"nocopy":              false,
+				"autoarchive":         false,
+				"allowexistingsource": false,
+				"version":             false},
+			stringFlags: map[string]string{
+				"user":          "",
+				"token":         "",
+				"linkfiles":     "keepInternalOnly",
+				"addattachment": "",
+				"addcaption":    ""},
+			intFlags: map[string]int{
+				"tapecopies": 0},
+			args: []string{"test"},
+		},
+		{
+			name: "Set all flags",
+			boolFlags: map[string]bool{
+				"ingest":              true,
+				"testenv":             true,
+				"devenv":              true,
+				"localenv":            true,
+				"tunnelenv":           true,
+				"noninteractive":      true,
+				"copy":                true,
+				"nocopy":              true,
+				"autoarchive":         true,
+				"allowexistingsource": true,
+				"version":             true},
+			stringFlags: map[string]string{
+				"user":          "usertest:passtest",
+				"token":         "token",
+				"linkfiles":     "somerandomstring",
+				"addattachment": "random attachment string",
+				"addcaption":    "a seemingly random caption"},
+			intFlags: map[string]int{
+				"tapecopies": 6571579},
+			args: []string{
+				"test",
+				"--ingest",
+				"--testenv",
+				"--devenv",
+				"--localenv",
+				"--tunnelenv",
+				"--noninteractive",
+				"--user",
+				"usertest:passtest",
+				"--token",
+				"token",
+				"--copy",
+				"--nocopy",
+				"--tapecopies",
+				"6571579",
+				"--autoarchive",
+				"--linkfiles",
+				"somerandomstring",
+				"--allowexistingsource",
+				"--addattachment",
+				"random attachment string",
+				"--addcaption",
+				"a seemingly random caption",
+				"--version",
+			},
+		},
 	}
 
-	os.Args = []string{"test"}
-	main()
-}
+	// running test cases
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			flag.CommandLine = flag.NewFlagSet(test.name, flag.ExitOnError)
+			datasetUtils.TestFlags = func(flags map[string]interface{}) {
+				passing := true
 
-func TestMainFlagSetting(t *testing.T) {
-	flag.CommandLine = flag.NewFlagSet("flag setting", flag.ExitOnError)
+				for bFlag := range test.boolFlags {
```

Comment:
why are the 3 types kept distinct? In the end, don't we only want to compare the values?

Reply:
I don't know if go lets programmers attempt to compare generic interface{} types. I could try doing that in order to combine the types, if you'd like!

Reply:
don't bother! thanks!